### PR TITLE
Copyright script update

### PR DESCRIPTION
**Pull request by kthui**

### CODE REVIEW

File: qa/common/check_copyright.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+# Copyright (c) 2020,21 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (aramesh7) [inline qa/common/check_copyright.py:1]: "2020-2021"

File: qa/common/check_copyright.py [hunk 2]
```diff
@@ -40,6 +43,7 @@
 
 single_re = re.compile(COPYRIGHT_YEAR_RE0)
 range_re = re.compile(COPYRIGHT_YEAR_RE1)
+new_range_re = re.compile(COPYRIGHT_YEAR_RE2)
```
Comment (aramesh7) [inline qa/common/check_copyright.py:46]: call this `affiliate_re` or something relevant to the change

File: qa/common/check_copyright.py [hunk 3]
```diff
@@ -105,34 +109,42 @@ def visit(path):
...
+                        elif len(year) == 2:  # 21
+                            years.append(int(year) + 2000)
+                        else:  # 21-23
+                            years.append(int(year[0:2]) + 2000)
+                            years.append(int(year[3:5]) + 2000)
+                else:
+                    print("copyright year is not recognized for " + path +
+                          ": " + line)
+                    return False
+        if years[0] > FLAGS.year:
```
Comment (aramesh7) [inline qa/common/check_copyright.py:139]: This checks that the start year is greater than the current year, and that the the years are increasing, but does it check whether the end year is greater than the current year?

File: qa/common/check_copyright.py [hunk 4]
```diff
@@ -105,34 +109,42 @@ def visit(path):
...
-            print("copyright end year greater than current year for " + path +
-                  ": " + line)
-            return False
-        if end_year < start_year:
-            print("copyright start year greater than end year for " + path +
+                copyright_row = line[len(prefix):]
+                if new_range_re.match(copyright_row):
+                    year_str = copyright_row.split("(c) ")[1].split(
+                        " NVIDIA ")[0]
+                    for year in year_str.split(","):
```
Comment (aramesh7) [inline qa/common/check_copyright.py:127]: Will this still work if we have `Copyright (c) 2020-2021 NVIDIA CORPORATION & AFFILIATES. All rights reserved.`?

**Review by aramesh7 — CHANGES_REQUESTED** [review]

File: qa/common/check_copyright.py (hunk 2 above)
Comment (kthui) [reply qa/common/check_copyright.py:46]: Yes, updated.

File: qa/common/check_copyright.py (hunk 3 above)
Comment (kthui) [reply qa/common/check_copyright.py:139]: Updated

File: qa/common/check_copyright.py (hunk 1 above)
Comment (kthui) [reply qa/common/check_copyright.py:1]: Will use `2020-21`

File: qa/common/check_copyright.py (hunk 4 above)
Comment (kthui) [reply qa/common/check_copyright.py:127]: Will be updated to support examples below:
`Copyright (c) 2018-20 NVIDIA CORPORATION & AFFILIATES. All rights reserved.`
`Copyright (c) 2018-19,21 NVIDIA CORPORATION & AFFILIATES. All rights reserved.`
`Copyright (c) 2018-19,21-22,24,26-28 NVIDIA CORPORATION & AFFILIATES. All rights reserved.`
`Copyright (c) 2020,22-23 NVIDIA CORPORATION & AFFILIATES. All rights reserved.`

**Comment by kthui** [comment]
Updated and rebased

**Review by aramesh7 — APPROVED** [review]

**Comment by kthui** [comment]
Rebased

**Review by aramesh7 — COMMENTED** [review]
LGTM. @CoderHam does this look good to you?

File: qa/common/check_copyright.py [hunk 5]
```diff
@@ -105,34 +109,49 @@ def visit(path):
...
+                copyright_row = line[len(prefix):]
+                if affiliate_re.match(copyright_row):
+                    year_str = copyright_row.split("(c) ")[1].split(
+                        " NVIDIA ")[0]
+                    for year in year_str.split(","):
+                        if len(year) == 4:  # 2021
+                            years.append(int(year))
+                        elif len(year) == 7:  # 2021-22
+                            years.append(int(year[0:3]))
+                            years.append(int(year[5:6]) + 2000)
```
Comment (CoderHam) [inline qa/common/check_copyright.py:132]: This should be 5:7

**Review by CoderHam — CHANGES_REQUESTED** [review]

**Comment by CoderHam** [comment]
@kthui I think there is just one small fix that I requested. Otherwise LGTM.

File: qa/common/check_copyright.py (hunk 5 above)
Comment (kthui) [reply qa/common/check_copyright.py:132]: Yes, this is wrong. Corrected

**Comment by kthui** [comment]
Currently, files that failed the copyright test are: 
`/.gitattributes`, `/.gitignore`, `/LICENSE`, `/VERSION`, `/.git/*`, `/examples/detailed_report.pdf` and `/examples/summary.pdf`.
They should be fine without passing the test.

**Review by CoderHam — APPROVED** [review]

**Comment by deadeyegoodwin** [comment]
@kthui  please apply this same change to server repo, file qa/common/check_copyright.py

**Comment by kthui** [comment]
@deadeyegoodwin In progress

**Comment by kthui** [comment]
Updated to support almost all formats, including those with or without ` (c)` and/or ` & AFFILIATES`, and the use of full/abbreviate year(s).

**Comment by kthui** [comment]
Currently, files that failed the copyright check script are:
`/.gitattributes`, `/.gitignore`, `/LICENSE`, `/VERSION`, `/.git/*`, `/examples/detailed_report.pdf` and `/examples/summary.pdf`.
They are the same as before, and should be fine.

**Comment by Tabrizian** [comment]
@kthui https://github.com/triton-inference-server/model_analyzer/blob/main/qa/common/check_copyright.py#L21 You should add the files that you want them to be skipped in the line I pasted.

**Comment by kthui** [comment]
I think there is a bug with how the `SKIP_PATHS` are checked:
If I call with `python3 ./qa/common/check_copyright.py -y 2021 VERSION`, the existing skip of `VERSION` works.
If I call with `python3 ./qa/common/check_copyright.py -y 2021 ./VERSION`, `VERSION` is not skipped.

I'll see if I can fix it, given the relative position of the check_copyright.py in the repo

**Comment by kthui** [comment]
```
> python3 ./qa/common/check_copyright.py -y 2021 .
> 
```

**Comment by kthui** [comment]
```
> python3 ./qa/common/check_copyright.py -y 2021 .
> 
```

**Comment by Tabrizian** [comment]
I see, that is probably why they were missed in the first place.

**Comment by kthui** [comment]
```
> python3 ./qa/common/check_copyright.py -y 2021 .
>
```

**Comment by kthui** [comment]
```
> python3 ./qa/common/check_copyright.py -y 2021 .
>
```

**Pull request merged**